HPCC-17152 Abort and improve error message if slaves don't respond

jakesmith opened the pull request:
Previously it only logged the error and stalled.
Now issues and exception and aborts the job and lists the slaves
that didn't respond that caused the timeout.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

Had to fake the lack of response from the slaves, by introducing fake delays.

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana committed:
HPCC-16964 Corrects inaccurate dali related default conf values
- No longer inserts optional elements/attributes
- Removes "outputBool" template and all references to that template,
   per component owner's request

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

afishbeck committed:
HPCC-17021 Improve result monitor query error handling

Capture and fail on exceptions returned by query being monitored.  We
don't want any ambiguity about the state of what is stored in cassandra.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

wangkx committed:
HPCC-14681 Fix "double free" memory in ESP esdl logging

In https://github.com/hpcc-systems/HPCC-Platform/pull/9610,
three set methods are used to pass log entry items. The set
methods call setown on those items without using LINK. The
LINKs are added in this fix.

Related functions are renamed to indicate the ownership
requirement.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

ghalliday committed:
HPCC-17141 Fix -ve memory leak processing bitfields

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

jakesmith committed:
HPCC-17152 Abort and improve error message if slaves don't respond

Previously it only logged the error and stalled.
Now issues and exception and aborts the job and lists the slaves
that didn't respond that caused the timeout.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
@mckellyln - please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 8ea15c2d52df9e6f5982b6a7cf7afea00aa9a5f8
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 40.788 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |86 |86 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

mckellyln commented on thorlcr/graph/thgraphmaster.cpp:
```diff
@@ -2529,14 +2529,31 @@ void CMasterGraph::getFinalProgress()
+                if (s>=n)
+                    break;
+            }
+            throw MakeGraphException(this, 0, "Timeout receiving final progress from slaves - these slaves failed to respond: %s", slaveList.str());
```
Is the test above for slaves who have responded already or or have not responded ?

jakesmith replied:
No, I have it wrong way around! reporting responding slaves not non-responders..
Will fix.

jakesmith replied:
In fact it was correct - it was supposed to list slaves who hadn't responded and did.
I've added a couple of comments.

But I also noticed a bug elsewhere, bitset was setting 1-based (slave 1 = bit 1), when wanted it to be 0-based. Fixed.

mckellyln replied:
If bitmask was 11001 (rightmost least significant) then would we skip s=3 scan ?

jakesmith replied:
It would, but that's intentional, don't want to scan from 3 (which was last scan result), want next scan be from 4.

mckellyln replied:
ok, yes I agree that is correct.

ghalliday committed:
HPCC-17148 Fix various minor warnings and unreachable code

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

jakesmith commented:
@mckellyln - please re-review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 902da37d5726a93c3cc8b610abe996c8410ca12e
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.87 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

Commit:
HPCC-17161 Add delete confirmation dialog

All details area should container confirmation dialog like the query areas do.

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexis.com>

ghalliday committed:
HPCC-17170 For regression tests split timing information into a separate file

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

wangkx committed:
HPCC-17151 Throw exception if Thor group not found for ZAP report

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx committed:
HPCC-16986 Remove default value of HasArchiveQuery in WsWorkunits

THe HasArchiveQuery is in struct ECLWorkunit which is used by
multiple WsWorkunits methods, such as WUInfo, WUQuery, etc. The
value is set in WUInfo, not in other methods. The default value
may give wrong information.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

Michael-Gardner committed:
HPCC-17123 Added descriptive error message and fixed init_thorslave call

Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

Commit:
HPCC-15693 Add dfuserver queue name selection

Whenever spraying in ECLWatch we should give the user the ability to select a different queue if necessary.

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexis.com>

mckellyln commented:
@jakesmith code looks good.

mckellyln commented:
@jakesmith code looks good.

RussWhitehead committed:
HPCC-17144 Strange code in getScopePermissions

getScopePermissions is allowing callers to pass in an invalid auditflags
argument which callers use to determine whether or not permissions are
available (security enabled).  Callers use this hack to first determine
if they should perform security checks.  This PR removes that peculiar
behavior and exposes the once private "permissionsavail" flag, which allows
callers to perform this check in a more intuitive manner

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

jakesmith commented:
@ghalliday - please merge.

ghalliday committed:
HPCC-17170 Remove unneeded addTimingToWorkunit option

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-17197 Stop AnnotationNormalizer creating inconsistent graphs

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

wangkx committed:
HPCC-14681 Remove ESP LoggingClient code

The code is not used.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

afishbeck committed:
HPCC-17021 Use more consistent naming for record layouts

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

ghalliday committed:
HPCC-17202 Improve code for short const sets of variable length strings

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Merge pull request #9674 from RussWhitehead/scopePerms640

HPCC-17144 Strange code in getScopePermissions

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday commented:
@jakesmith code looks good.  Should it target 6.4.0?

jakesmith commented:
Could do, I'll retarget.

richardkchapman committed:
Merge pull request #9628 from rpastrana/HPCC-16964-DaliXSLCleanup

HPCC-16964 Corrects inaccurate dali related default conf values

Reviewed-By: Jake Smith <jake.smith@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9604 from afishbeck/esdlMonExceptions2

HPCC-17021 Improve result monitor query error handling

Reviewed-By: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9641 from wangkx/h14681b

HPCC-14681 Fix "double free" memory in ESP esdl logging

Reviewed-By: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9679 from wangkx/h14681_remove_esp_loggingclient

HPCC-14681 Remove ESP LoggingClient code

Reviewed-By: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9662 from miguelvazq/HPCC-17161

HPCC-17161 Add delete confirmation dialog

Reviewed-By: Gordon Smith <gordon.smith@lexisnexis.com>
Reviewed-By: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9664 from ghalliday/issue17170

HPCC-17170 For regression tests split timing information into a separate file

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9677 from ghalliday/issue17197

HPCC-17197 Stop AnnotationNormalizer creating inconsistent graphs

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9673 from miguelvazq/HPCC-15963

HPCC-15693 Add dfuserver queue name selection

Reviewed-By: Gordon Smith <gordon.smith@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9671 from wangkx/h16986

HPCC-16986 Remove default value of HasArchiveQuery in WsWorkunits

Reviewed-By: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9669 from wangkx/h17151

HPCC-17151 Throw exception if Thor group not found for ZAP report

Reviewed-By: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9672 from Michael-Gardner/HPCC-17123

HPCC-17123 Added descriptive error message and fixed init_thorslave call

Reviewed-By: Xiaoming Wang <xiaoming.wang@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9657 from ghalliday/issue17148

HPCC-17148 Fix various minor warnings and unreachable code

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9652 from ghalliday/issue17141

HPCC-17141 Fix -ve memory leak processing bitfields

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #9683 from ghalliday/issue17202

HPCC-17202 Improve code for short const sets of variable length strings

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

jakesmith commented:
Closing to re-target 6.4

jakesmith commented:
@ghalliday - changed target of existing PR (didn't know you could do that until now)

Please merge.

richardkchapman commented:
@jakesmith you can, but you often (as in this case) need to also rebase and repush (with force) the source branch

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 292437de98c779e1fb71b27e67bcc80747448321
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.772 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
temp. closing
